### PR TITLE
Cobalt Platform service on Linux b/332929552

### PR DESCRIPTION
**Pull request by yzzhang11**
Create Cobalt platform service registry on Linux;
Decouple Soft mic platform service from Cobalt extension platform service;

b/332929552

### CODE REVIEW

File: starboard/linux/shared/soft_mic_platform_service.cc [hunk 1]
```diff
@@ -48,51 +43,27 @@ const char kJSONFalse[] = "false";
...
-  // with free (), since new[] will map to malloc()
-  // in an Evergreen build.
-  if (is_evergreen) {
-    free((void*)name);  // NOLINT
-  } else {
-    delete[] name;
-  }
+  SB_LOG(INFO) << "Open() service created: " << kSoftMicPlatformServiceName;
+  SoftMicPlatformServiceImpl* soft_mic_impl =
+      new SoftMicPlatformServiceImpl({context, receive_callback});
```
Comment (y4vor) [inline starboard/linux/shared/soft_mic_platform_service.cc:57]: Our old clang 3.9 builders doesn't know how to create a ctor with the ilitialaliztion list. Please add a ctor to the type.
https://github.com/youtube/cobalt/actions/runs/8651949051/job/23723874730?pr=2896#step:6:15616

File: starboard/linux/shared/soft_mic_platform_service.h [hunk 2]
```diff
@@ -19,7 +19,9 @@
 namespace starboard {
 namespace shared {
 
-const void* GetPlatformServiceApi();
+#define kSoftMicPlatformServiceName "com.google.youtube.tv.SoftMic"
```
[the remark this answers is not in the document]
Comment (yzzhang11) [reply starboard/linux/shared/soft_mic_platform_service.h]: Make sense. Done.

File: starboard/linux/shared/platform_service.cc [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+  if (!Has(name)) {
+    SB_LOG(ERROR) << "Open() failed in service: " << name;
+    service = kCobaltExtensionPlatformServiceInvalid;
+  } else {
+    const CobaltPlatformServiceApi* api =
+        reinterpret_cast<const CobaltPlatformServiceApi*>(
+            platform_service_registry.at(name)());
+
+    char* service_name = new char[strlen(name) + 1];
```
[the remark this answers is not in the document]
Comment (yzzhang11) [reply starboard/linux/shared/platform_service.cc]: Done.

File: starboard/linux/shared/soft_mic_platform_service.cc (hunk 1 above)
Comment (yzzhang11) [reply starboard/linux/shared/soft_mic_platform_service.cc]: Done.

File: starboard/linux/shared/platform_service.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                uint64_t length,
+                uint64_t* output_length,
+                bool* invalid_state);
+} CobaltPlatformServiceApi;
+
+typedef struct CobaltExtensionPlatformServicePrivate {
+  void* context;
+  ReceiveMessageCallback receive_callback;
+  const char* name;
+  PlatformServiceImpl* platform_service_impl;
```
[the remark this answers is not in the document]
Comment (yzzhang11) [reply starboard/linux/shared/platform_service.h]: Make sense. Done.

**Review by y4vor — APPROVED** [review]

**Pull request merged**